<h1>Финальны проект SQL<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных:" data-toc-modified-id="Описание-данных:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных:</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Задания</a></span></li><li><span><a href="#Выгрузка-данных" data-toc-modified-id="Выгрузка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выгрузка данных</a></span><ul class="toc-item"><li><span><a href="#Ознакомление-с-данными" data-toc-modified-id="Ознакомление-с-данными-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Ознакомление с данными</a></span></li></ul></li><li><span><a href="#Решение-задач" data-toc-modified-id="Решение-задач-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Решение задач</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li><li><span><a href="#Общие-выводы" data-toc-modified-id="Общие-выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общие выводы</a></span></li></ul></div>

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных:

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


## Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

## Выгрузка данных


In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

con=engine.connect()

pd.set_option('display.max_columns', 100)

### Ознакомление с данными

In [2]:
#Напишем функцию для вывода запросов SQL в pandas
def use_request(request):
    return pd.io.sql.read_sql(sql=text(request), con = con)

In [3]:
#создаем список таблиц
df_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']

#Через цикл выводим их содержимое
for i in df_list:
    show = '''SELECT * FROM {} '''.format(i)
    display(use_request(show).head())
    display(use_request(show).info())
    display(use_request(show).isnull().sum())
    print(f'Количество дубликатов равно {use_request(show).duplicated().sum()}')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

Количество дубликатов равно 0


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id    0
author       0
dtype: int64

Количество дубликатов равно 0


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id    0
publisher       0
dtype: int64

Количество дубликатов равно 0


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

Количество дубликатов равно 0


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id    0
book_id      0
username     0
text         0
dtype: int64

Количество дубликатов равно 0


В целом данные хорошие, пропусков нет как и дублей тоже,формат данных подходящий, придраться можно лишь к формату даты в books.publication_date  сейчас там `object`.
В нашем распоряжении данные о:

* 1000 книг
* 636 авторах 
* 2793 обзоров
* 340 издательств
* 6456 оценок 
* 2793  обзоров

## Решение задач

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [4]:
#пишем запрос
book_cout = ''' SELECT COUNT(DISTINCT book_id)
                FROM books 
                WHERE publication_date::date >= '2000-01-02';
            '''
use_request(book_cout)

count
0    819

Всего книг вышедших после 1 явнаря 2000 года 819 штук. Если идти от обратного, то 181 книга вышло до

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [5]:
#проводим группировку по книге и названию, считая количество обзоров из `reviews`, а подтягивая данные из `ratings` считаем AVG
review_cout = ''' SELECT DISTINCT (b.book_id),
                         b.title,
                         COUNT(rev.review_id) count_review,
                         AVG(rat.rating) avg_rating
                         
                  FROM books b
                  LEFT JOIN reviews rev ON b.book_id = rev.book_id
                  LEFT JOIN ratings rat ON rev.book_id = rat.book_id
                  GROUP BY 1,2
                  ORDER BY avg_rating DESC
                  LIMIT 15;
               '''
use_request(review_cout)

book_id                                              title  count_review  \
0        83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
1       191  Disney's Beauty and the Beast (A Little Golden...             0   
2       221                          Essential Tales and Poems             0   
3       387                               Leonardo's Notebooks             0   
4       672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
5       808                            The Natural Way to Draw             0   
6        17                      A Dirty Job (Grim Reaper  #1)            16   
7        20              A Fistful of Charms (The Hollows  #4)             4   
8        55          A Woman of Substance (Emma Harte Saga #1)             4   
9        57                    Act of Treason (Mitch Rapp  #9)             4   
10       62                                      Alas  Babylon             4   
11       76                                        Angels Fall             4   
12       86      Arrows of the Queen (Heralds of Valdemar  #1)             4   
13      136  Captivating: Unveiling the Mystery of a Woman'...             4   
14      169  Crucial Conversations: Tools for Talking When ...             4   

    avg_rating  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  
5          NaN  
6          5.0  
7          5.0  
8          5.0  
9          5.0  
10         5.0  
11         5.0  
12         5.0  
13         5.0  
14         5.0

6 книг не получили оценок и достаточно большое количество со среднийм рейтингом 5. Однако, можеты быть, это одна единсвенная оценка книги). В данном списке я скоре пригляделся к книге с высоким количество обзоров `A Dirty Job (Grim Reaper #1)`

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [6]:
#опять же данные начинаем тянут от книги подтягивая строки из данных об издательствах

publisher_cout = ''' SELECT publisher,
                         COUNT(DISTINCT b.book_id)
                  FROM books b 
                  LEFT JOIN publishers p ON b.publisher_id  = p.publisher_id 
                  WHERE num_pages >50
                  GROUP BY 1
                  ORDER BY 2 DESC
                  LIMIT 1;
              '''
use_request(publisher_cout)

publisher  count
0  Penguin Books     42

42 книги выпустило издательсво `Penguin Books` и это наибольший результат	

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [7]:
#В rating определяем книги, которые получили 50 и более оценока
#В profiles подтигиваем назвавние авторов, подсчитываем количество книг и суммарный бал
# В основной вывод определяем имя автора и произведенеие суммарного бала, на количество оценок

max_rating  = ''' WITH 
                  rating AS ( SELECT    book_id,
                                        COUNT(rating_id),
                                        AVG(rating) avg_rating
                              
                              FROM ratings
                              GROUP BY 1
                              HAVING COUNT(rating_id ) >= 50
                              ORDER BY  3 DESC
                             )
                             
                  SELECT author,
                         sum/count avg_rait
                  
                  FROM ( SELECT author,
                               COUNT(r.book_id),
                               SUM(r.avg_rating)
                         
                         FROM rating r
                         LEFT JOIN books b ON r.book_id = b.book_id
                         LEFT JOIN authors a ON b.author_id = a.author_id
                         GROUP BY 1
                         
                       ) AS profiles
                       
                       
                  ORDER BY 2 DESC
                  LIMIT 1;
              '''
use_request(max_rating)

author  avg_rait
0  J.K. Rowling/Mary GrandPré  4.283844

Автор с самой высокой среденей оценкой J.K. Rowling/Mary GrandPré

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [8]:
#сперва вычисляем пользователей которые поавтили более 48 оценок
#затем считаем количество пользователей им сумм обзоров

count_review  = ''' 
                                     
                WITH
                super_profiles AS (SELECT profiles.username,
                                          COUNT(review_id) count_review
                  
                                    FROM (SELECT username,
                                          COUNT(rating_id ) count_rating
                                   
                                          FROM ratings 
                                          GROUP BY 1
                                          HAVING COUNT(rating_id)>48)  as profiles
                  
                                          LEFT JOIN reviews rev ON profiles.username = rev.username
                                          GROUP BY 1
                                    ) 
                
                  SELECT SUM(count_review)/COUNT(username) avg_reviews 
                  FROM super_profiles;
                
              '''
use_request(count_review)

avg_reviews
0         24.0

Среднее количество обзоров от пользователей, поставивших более 48 оценок, ровняется  24 шт

## Общие выводы
 * Всего книг вышедших после 1 явнаря 2000 года 819 штук. Если идти от обратного, то 181 книга вышло до
 
 
 * 6 книг не получили оценок и достаточно большое количество со среднийм рейтингом 5. Однако, можеты быть, это одна единственная оценка книги). В данном списке я скоре пригляделся к книге с высоким количество обзоров `A Dirty Job (Grim Reaper #1)`
 
 
 * 42 книги выпустило издательсво `Penguin Books` и это наибольший результат	
 
 
 * Автор с самой высокой среденей оценкой J.K. Rowling/Mary GrandPré
 
 
 * Среднее количество обзоров от пользователей, поставивших более 48 оценок, ровняется  24 шт